filestorage.go

```diff
@@ -199,27 +214,29 @@ createLock:
 			}
 
 			// if lock is not stale, wait upon it
-			return fw, nil
+			fw.Wait()
+			return nil
 		}
 
 		// otherwise, this was some unexpected error
-		return nil, err
+		return err
 	}
 	lf.Close()
 
-	// looks like we get the lock
-	fw.wg.Add(1)
-	fileStorageNameLocks[key] = fw
-
-	return nil, nil
+	// cool, we got the lock right away
+	return nil
 }
 
 // Unlock releases the lock for name.
 func (fs FileStorage) Unlock(key string) error {
```

Comment:
typing again after closing the tab....

I think the lockDir removal might be unsafe.
I don't know about on Windows, but on `*nix`, `os.Remove()` should fail if the directory isn't empty, so you don't need to open and read its contents - which is racy because another process might come along and change things between us reading the filenames and calling `os.Remove()`.

In either case, I think the whole parent directory removal is possibly racy:

* we remove the file
* other process proceeds since the file no longer exists
* other process calls mkdirall
* we proceed to remove the directory
* other process fails to create file in the now non-existent directory

I'm not sure you need to remove it, though, because it's always the same directory and it's just going to get re-created anyway.